src/appservice/Appservice.ts

```diff
@@ -986,18 +996,28 @@ export class Appservice extends EventEmitter {
         }
 
         let responded = false;
-        this.emit("query.key", req.body, async (result: MSC3984KeyQueryResponse | Promise<MSC3984KeyQueryResponse> | undefined | null) => {
-            if ((result as any)?.then) result = await result;
-            if (!result) {
-                res.status(404).json({ errcode: "M_UNRECOGNIZED", error: "Endpoint not implemented" });
-                responded = true;
-                return;
-            }
+        this.emit("query.key", req.body, (result: MSC3984KeyQueryResponse | Promise<MSC3984KeyQueryResponse> | undefined | null) => {
+            responded = true;
 
-            // Implementation note: we could probably query the device keys from our storage if we wanted to.
+            const handleResult = (result2: MSC3984KeyQueryResponse) => {
+                if (!result2) {
+                    res.status(404).json({ errcode: "M_UNRECOGNIZED", error: "Endpoint not implemented" });
+                    return;
+                }
 
-            res.status(200).json(result);
-            responded = true;
+                // Implementation note: we could probably query the device keys from our storage if we wanted to.
+
+                res.status(200).json(result2);
+            };
+
+            if ((result as Promise<MSC3984KeyQueryResponse>)?.then) {
+                (result as Promise<MSC3984KeyQueryResponse>).then(r => handleResult(r)).catch(e => {
+                    LogService.error("Appservice", "Error handling key query API", e);
+                    res.status(500).json({ errcode: "M_UNKNOWN" });
```

Comment:
Could we at least emit an `error` so the consumer has something to go off "Error handling key claim".